pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,526 @@ package service
+// And it returns the rebalancer implemenation or any error occurred.
+func NewRebalancer(opts ...RebalancerOption) (Rebalancer, error) {
 	r := new(rebalancer)
-	for _, opt := range append(defaultOpts, opts...) {
+
+	for _, opt := range append(defaultRebalancerOpts, opts...) {
 		if err := opt(r); err != nil {
 			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 		}
 	}
+
+	job, err := job.New(
+		job.WithControllerName("job rebalancer"),
+		job.WithNamespaces(r.jobNamespace),
+		job.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		job.WithOnReconcileFunc(func(jobList map[string][]job.Job) {
+			log.Debugf("[reconcile Job] length Joblist: %d", len(jobList))
+			jobs, ok := jobList[r.jobName]
+			if ok {
+				r.jobs.Store(jobs)
+			} else {
+				r.jobs.Store(make([]job.Job, 0))
+				log.Infof("job not found: %s", r.jobName)
+			}
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cm, err := configmap.New(
+		configmap.WithControllerName("configmap rebalancer"),
+		configmap.WithNamespaces(r.jobNamespace),
+		configmap.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		configmap.WithOnReconcileFunc(func(configmapList map[string][]configmap.ConfigMap) {
+			configmaps, ok := configmapList[r.jobNamespace]
+			if ok {
+				for _, cm := range configmaps {
+					if cm.Name == r.jobConfigmapName {
+						if tmpl, ok := cm.Data[r.jobTemplateKey]; ok {
+							r.jobTemplate.Store(tmpl)
+						} else {
+							log.Infof("job template not found: %s", r.jobTemplateKey)
+						}
+						break
+					}
+				}
+			} else {
+				log.Infof("configmap not found: %s", r.jobConfigmapName)
+			}
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var rc k8s.ResourceController
+	switch r.agentResourceType {
+	case config.STATEFULSET:
+		rc, err = statefulset.New(
+			statefulset.WithControllerName("statefulset rebalancer"),
+			statefulset.WithNamespaces(r.agentNamespace),
+			statefulset.WithOnErrorFunc(func(err error) {
+				log.Error(err)
+			}),
+			statefulset.WithOnReconcileFunc(func(statefulSetList map[string][]statefulset.StatefulSet) {
+				log.Debugf("[reconcile StatefulSet] length StatefulSet[%s]: %d", r.agentName, len(statefulSetList))
+				sss, ok := statefulSetList[r.agentName]
+				if ok {
+					log.Debugf("[reconcile StatefulSet] StatefulSet[%s]: desired replica: %d, current replica: %d", r.agentName, *sss[0].Spec.Replicas, sss[0].Status.Replicas)
+					if len(sss) == 1 {
+						pss, ok := r.statefulSets.Load().(statefulset.StatefulSet)
+						if ok && *sss[0].Spec.Replicas < *pss.Spec.Replicas {
+							jobTpl, err := r.genJobTpl()
+							if err != nil {
+								log.Errorf("[recovery] error generating job template: %s", err.Error())
+							} else {
+								for i := int(*pss.Spec.Replicas); i > int(*sss[0].Spec.Replicas); i-- {
+									name := r.agentName + "-" + strconv.Itoa(i-1)
+									log.Debugf("[recovery] creating job for pod %s", name)
+									ctx := context.TODO()
```

Comment:
context must be passed in an reconcile loop.

Reply:
thank you.
I fixed: https://github.com/vdaas/vald/pull/1187/commits/ad9415a202d97434c3c1c1a98a4d8d9e3897a39c